File: day6/elfsignal.py
#!/usr/bin/env python

def unique_run(s, l):
        for i in range(len(line) - l):
            s = set(line[i:i+l])
            if len(s) == l:
                return i+l
        return None


with open('input', 'r') as _in:
    for line in _in:
        line = line.strip()
        print("signal after", unique_run(line, 4))        
        print("message after", unique_run(line, 14))

File: day6/elfsignal.clj
(defn unique-run
  [signal length]
  (loop [remainder signal pos 0]
    (if (< (count remainder) length)
      0
      (if (= length (count (set (take length remainder))))
        (+ pos 4)
        (recur (rest remainder) (inc pos))))))

(defn parsefile
  "Get a line from a file and check if for signal and message"
  [filename]
  (with-open [rdr (clojure.java.io/reader filename)]
    (let [line (first (line-seq rdr))]
      (do
        (println (unique-run line 4))
        (println (unique-run line 14))))))

(parsefile "input")
